File: examples/web-nodejs/src/backend/utils/string.js
const crypto = require("node:crypto").webcrypto;

const randomString = (chars, length) => {
  const randomArray = new Uint8Array(length);
  crypto.getRandomValues(randomArray);
  return Array.from({length}, (_, i) => chars[randomArray[i] % chars.length]).join("");
};

module.exports = { randomString };

File: examples/android-flutter/lib/utils/string.dart
import 'dart:math';
import 'dart:convert';
import 'package:crypto/crypto.dart';

String multiline(String str) {
  return str.replaceAll(RegExp(r'\n(\s{2,})+'), ' ').trim();
}

String randomString(List<String> chars, int len) {
  var r = Random.secure();

  return List<String>.generate(len, (_) => chars[r.nextInt(chars.length)]).join();
}

Digest digestMessage(String message) {
  var bytes = utf8.encode(message);
  return sha256.convert(bytes);
}

String base64urlencode(List<int> bytes) {
  // Generates an unpadded base64 encoded URL string, see:
  // https://datatracker.ietf.org/doc/html/rfc7636#appendix-A
  String encoded = base64.encode(bytes);
  return encoded.split('=')[0].replaceAll('+', '-').replaceAll('/', '_');
}
